### ISSUE TITLE
Assertion error during instruction selection

### ISSUE THREAD

**Issue by dylanmckay** [avr-llvm/llvm#100]
Message:
```
Assertion error: "#operands for dag node doesn't match .td file!"
```

Command line:
```
llc -filetype=asm -march=avr -mcpu=atmega328p fourier.ll -o fourier.s
```

Reduced test case:
``` llvm
target datalayout = "e-p:16:8:8-i8:8:8-i16:8:8-i32:8:8-i64:8:8-f32:8:8-f64:8:8-n8"
target triple = "avr"

define i32 @main(i32 %x, i32 %in_min, i32 %in_max, i32 %out_min, i32 %out_max) {
entry:
  %sub = sub nsw i32 %x, %in_min
  %sub1 = sub nsw i32 %out_max, %out_min
  %mul = mul nsw i32 %sub1, %sub
  %sub2 = sub nsw i32 %in_max, %in_min
  %div = sdiv i32 %mul, %sub2
  %add = add nsw i32 %div, %out_min
  ret i32 %add
}
```

Program output:
```
llc: /home/dylan/projects/avr-llvm/llvm/lib/CodeGen/SelectionDAG/InstrEmitter.cpp:780: void llvm::InstrEmitter::EmitMachineNode(llvm::SDNode*, bool, bool, llvm::DenseMap<llvm::SDValue, unsigned int>&): Assertion `NumMIOperands >= II.getNumOperands() && NumMIOperands <= II.getNumOperands() + II.getNumImplicitDefs() + NumImpUses && "#operands for dag node doesn't match .td file!"' failed.
#0 0x132b69d llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/dylan/projects/avr-llvm/llvm/lib/Support/Unix/Signals.inc:437:0
#1 0x132ba1a PrintStackTraceSignalHandler(void*) /home/dylan/projects/avr-llvm/llvm/lib/Support/Unix/Signals.inc:495:0
#2 0x132a5f1 SignalHandler(int) /home/dylan/projects/avr-llvm/llvm/lib/Support/Unix/Signals.inc:210:0
#3 0x7fca0b666660 __restore_rt (/usr/lib/libpthread.so.0+0x10660)
#4 0x7fca0a835528 __GI_raise (/usr/lib/libc.so.6+0x33528)
#5 0x7fca0a83693a __GI_abort (/usr/lib/libc.so.6+0x3493a)
#6 0x7fca0a82e3a7 __assert_fail_base (/usr/lib/libc.so.6+0x2c3a7)
#7 0x7fca0a82e452 (/usr/lib/libc.so.6+0x2c452)
#8 0x1229460 llvm::InstrEmitter::EmitMachineNode(llvm::SDNode*, bool, bool, llvm::DenseMap<llvm::SDValue, unsigned int, llvm::DenseMapInfo<llvm::SDValue>, llvm::detail::DenseMapPair<llvm::SDValue, unsigned int> >&) /home/dylan/projects/avr-llvm/llvm/lib/CodeGen/SelectionDAG/InstrEmitter.cpp:784:0
#9 0x10befef llvm::InstrEmitter::EmitNode(llvm::SDNode*, bool, bool, llvm::DenseMap<llvm::SDValue, unsigned int, llvm::DenseMapInfo<llvm::SDValue>, llvm::detail::DenseMapPair<llvm::SDValue, unsigned int> >&) /home/dylan/projects/avr-llvm/llvm/lib/CodeGen/SelectionDAG/InstrEmitter.h:124:0
#10 0x10d6019 llvm::ScheduleDAGSDNodes::EmitSchedule(llvm::MachineBasicBlock::bundle_iterator<llvm::MachineInstr, llvm::ilist_iterator<llvm::MachineInstr> >&) /home/dylan/projects/avr-llvm/llvm/lib/CodeGen/SelectionDAG/ScheduleDAGSDNodes.cpp:850:0
#11 0x1172ff2 llvm::SelectionDAGISel::CodeGenAndEmitDAG() /home/dylan/projects/avr-llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:819:0
#12 0x1171ad2 llvm::SelectionDAGISel::SelectBasicBlock(llvm::ilist_iterator<llvm::Instruction const>, llvm::ilist_iterator<llvm::Instruction const>, bool&) /home/dylan/projects/avr-llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:621:0
#13 0x1174d6a llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) /home/dylan/projects/avr-llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1316:0
#14 0x1170c7e llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) /home/dylan/projects/avr-llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:465:0
#15 0x9f804f llvm::AVRDAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) /home/dylan/projects/avr-llvm/llvm/lib/Target/AVR/AVRISelDAGToDAG.cpp:72:0
#16 0xbd4487 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /home/dylan/projects/avr-llvm/llvm/lib/CodeGen/MachineFunctionPass.cpp:41:0
#17 0xf8a019 llvm::FPPassManager::runOnFunction(llvm::Function&) /home/dylan/projects/avr-llvm/llvm/lib/IR/LegacyPassManager.cpp:1520:0
#18 0xf8a178 llvm::FPPassManager::runOnModule(llvm::Module&) /home/dylan/projects/avr-llvm/llvm/lib/IR/LegacyPassManager.cpp:1540:0
#19 0xf8a4a7 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /home/dylan/projects/avr-llvm/llvm/lib/IR/LegacyPassManager.cpp:1596:0
#20 0xf8ab08 llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/dylan/projects/avr-llvm/llvm/lib/IR/LegacyPassManager.cpp:1698:0
#21 0xf8ad0d llvm::legacy::PassManager::run(llvm::Module&) /home/dylan/projects/avr-llvm/llvm/lib/IR/LegacyPassManager.cpp:1730:0
#22 0x9c9cd5 compileModule(char**, llvm::LLVMContext&) /home/dylan/projects/avr-llvm/llvm/tools/llc/llc.cpp:367:0
#23 0x9c8e4d main /home/dylan/projects/avr-llvm/llvm/tools/llc/llc.cpp:204:0
#24 0x7fca0a822790 __libc_start_main (/usr/lib/libc.so.6+0x20790)
#25 0x9c7c59 _start (/home/dylan/projects/builds/avr-llvm/llvm/bin/llc+0x9c7c59)
Stack dump:
0.      Program arguments: ./llc -march=avr -mcpu=atmega328p -filetype=obj -o a.out asdf.ll 
1.      Running pass 'Function Pass Manager' on module 'asdf.ll'.
2.      Running pass 'AVR DAG->DAG Instruction Selection' on function '@main'
[1]    19129 abort (core dumped)  ./llc -march=avr -mcpu=atmega328p -filetype=obj -o a.out asdf.ll

```

Tested with 4353c50e5e2b358c581bbe2ff14b3d32a8f8d5b8.

**Comment by agnat** [comment]
Additional info: The (pseudo) opcode is `STDWPtrQRr` and it's an off-by-one... of course. `II.getNumOperands()` is three, which sounds correct. But `NumMIOperands` is four. The rest of the terms is zero.

So, it seems like there is a superfluous operand floating around. @dylanmckay, since you touched that area earlier... does that ring a bell?

**Comment by dylanmckay** [comment]
That instruction wasn't touched with my fix of the LDD/STD instructions that I think you're referring to.

The last time this instruction was modified was when I cleaned up the whole file, placing related instructions next to each other and instructions with common fields (isLoad,..) in blocks.

It is possible that the instruction lost a property along the way. Specifically I'm thinking of the "constraints" property, where "$srx=$dst" was not copied exactly and LLVM failed to recognize two always equal operands as one.

Just speculation from memory though, I'll check when I get home.

**Comment by dylanmckay** [comment]
Note that this error occurs while compiling two Arduino core files - `WMath.cpp` and `Print.cpp`.

* With `WMath.cpp`, the operand corresponds to `STDPtrQRr`
* With `Print.cpp`, the operand corresponds to `LDDWRdPtrQ`

**Comment by dylanmckay** [comment]
And with [`fourier.ll](https://raw.githubusercontent.com/avr-llvm/examples/master/fourier/fourier.ll), the opcode correspond to `STDWPtrQRr`.

**Comment by dylanmckay** [comment]
Fixed in af88a32108b0f11ee467cf2af5e282930bc61b3e.

The cause was that a long time ago I removed the `memri` custom operand type, thinking it was redundant and could be expressed in pure TableGen. It could - more or less.

The problem was that `reg+imm` addresses were still using a custom `ComplexPattern` (named `addr`) for instruction selection, which wasn't aware of this.

I decided the backend would be a lot cleaner if the `memri` operand kind was added back.
